### ISSUE TITLE
tlog client : buffer tlog data in case of failed server

### ISSUE THREAD

**Issue by iwanbk** [zero-os/0-Disk#605]
To avoid vdisk become failed when tlogservers are failed, tlogclient might need to buffer transaction data.

The buffering could be done in:
- memory, as first buffer place, until some amount of transactions
- disk, when the `memory` is considered `full`, also need to be limited in size

Of course we can't do buffering forever because the disk will be full after some amount of time.
But by that time, orchestrator should bring the tlogserver back.

If orchestrator can't do this, then the environment might be indeed in corrupt state.

**Comment by GlenDC** [comment]
Caching is never easy. The longer we keep things in memory, the bigger the impact of a crash would become. Writing stuff to disk is also not possible as I'm fairly sure that in the current architecture we can't rely on the fact that our servers have disks at all (even though they have). What I mean is that I'm fairly sure we don't want to store anything on them. What perhaps we could do is use some ardb servers as caching servers. But even so, that does make things more complicated as there are more moving pieces.

**Comment by iwanbk** [comment]
>  Writing stuff to disk is also not possible

I'm sure i see that nbd will use disk at #606

**Comment by GlenDC** [comment]
> I'm sure i see that nbd will use disk at #606

Not it will not. And if that PR does do that, than it will definitely not be merged in at all. Not that I was planning to merge it in either way.

**Comment by zaibon** [comment]
> Writing stuff to disk is also not possible as I'm fairly sure that in the current architecture we can't rely on the fact that our servers have disks at all 

True, I don't think we can rely on local disk for anything. Sometime there will be no disk at all and most of the time all disk will be used by storage clusters.

Can't the tlogclient keep a record of what it failed to send to the tlog while the tlog server is unreachable and as soon as the tlog comes back alive it send the required transaction.
Keeping this information shouldn't take too much memory right ?
And if the tlog server never comes back, then yes we'll have a problem but then that means that something super serious happens and there is not much we can do about this.

**Comment by iwanbk** [comment]
> Can't the tlogclient keep a record of what it failed to send to the tlog while the tlog server is unreachable and as soon as the tlog comes back alive it send the required transaction.
> Keeping this information shouldn't take too much memory right ?

Yeah, we could.
And it is also in my proposal (see issue description) : first using the memory and then use the disk.

So, memory is enough, agree

**Comment by iwanbk** [comment]
Moved to `beta-1` from `roadmap` milestone because it might relate to #541, which is what we need to do in case of tlogserver failed.
